driver/main.c

```diff
@@ -50,25 +47,72 @@ typedef struct {
   PetscReal xlc, ylc;      // cell centroid coordinates
   PetscReal cellsize;      // dx = dy of cell
 
-  PetscInt  *data2mesh_idx;
-  PetscReal *data_xc, *data_yc;
-  PetscReal *mesh_xc, *mesh_yc;
-  PetscInt   mesh_ncells_local;
+  PetscInt   mesh_ncells_local;  // number of cells in RDycore meshes
+  PetscInt  *data2mesh_idx;      // for each RDycore cell, the index of the data in the raster dataset
+  PetscReal *data_xc, *data_yc;  // x and y coordinates of data in the raster dataset
+  PetscReal *mesh_xc, *mesh_yc;  // x and y coordinates of RDycore cells
+
+  PetscBool output_map;  // if true, writes the mapping between the RDycore cells and the dataset
+} RasterDataset;
+
+typedef struct {
+  char dir[PETSC_MAX_PATH_LEN];
+  char file[PETSC_MAX_PATH_LEN];
+  char mesh_file[PETSC_MAX_PATH_LEN];
+
+  PetscReal dtime_in_hour;
+  PetscInt  ndata_file;
+
+  struct tm start_date, current_date;  // start and current date for rainfall dataset
+
+  // binary data
+  Vec          data_vec;
+  PetscScalar *data_ptr;
+
+  PetscInt ndata;
+  PetscInt stride;
+
+  PetscInt   mesh_nelements;     // number of cells or boundary edges in RDycore mesh
+  PetscInt  *data2mesh_idx;      // for each RDycore element (cells or boundary edges), the index of the data in the unstructured dataset
+  PetscReal *data_xc, *data_yc;  // x and y coordinates of data
+  PetscReal *mesh_xc, *mesh_yc;  // x and y coordinates of RDycore elments
+
+  PetscBool output_map;  // if true, writes the mapping between the RDycore cells and the dataset
 
-} HeterogeneousRainData;
+} UnstructuredDataset;
 
 typedef struct {
-  RainType              type;
-  ConstantRainData      constant;
-  HomogeneousRainData   homogeneous;
-  HeterogeneousRainData heterogeneous;
-} Rain;
-
-// open a Vec that contains data in the following format:
-//
-// time_1 value_1
-// time_2 value_2
-// time_3 value_3
+  DatasetType         type;
```

Comment:
We should discuss union data types when you're bored and have a moment. They are a more formal mechanism for having a "variant" type (in other words: a type that holds exactly one value belonging to a set of different possible types). But this is fine for now.